internal/jobcontainers/logon.go

```diff
@@ -0,0 +1,67 @@
+package jobcontainers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+)
+
+// Takes in a DOMAIN\Username or just Username and will return a token
+// for the account if successful.
+func processToken(user string) (windows.Token, error) {
+	var (
+		domain   string
+		userName string
+		token    windows.Token
+	)
+
+	split := strings.Split(user, "\\")
+	if len(split) == 2 {
+		domain = split[0]
+		userName = split[1]
+	} else if len(split) == 1 {
+		userName = split[0]
+	} else {
+		return token, fmt.Errorf("invalid user string `%s`", user)
+	}
+
+	// If empty, ContainerUser or ContainerAdministrator just let it inherit the token
+	// from whatever is used to launch it (containerd-shim etc). Regular container images
+	// are usable for job containers so we need to handle the ContainerUser and ContainerAdministrator
+	// cases.
+	if user == "" || user == "ContainerUser" || user == "ContainerAdministrator" {
```

Comment:
In effect this will default to running as SYSTEM in most cases. Should we instead default to a user account?

Reply:
Perhaps this should also be a config option.

Reply:
The question is, what user account?

Reply:
And config option to decide whether to default to just inheriting the parents token (in this case running as SYSTEM) or starting as <less_privileged_user>?

Reply:
I think running as Administrator is probably closer to what a customer would expect for a privileged container. Not sure if there are cases where that account is not available. Did we say anything in the KEP about the user it would run as?

I was thinking a config option to specify what user the container should run as. Otherwise it defaults to something.

Reply:
Would that config option just be https://github.com/kubernetes/cri-api/blob/2417d2c7e28a67dd9cec59c2d37ce3a1669f54f3/pkg/apis/runtime/v1alpha2/api.proto#L701 ?

Reply:
And if it's omitted we just default to Administrator/inheriting from shim/something else

Reply:
There are really two points here:
- What should the default fallback be if no user is specified or the user does not exist?
- Should we have a config option to change the default fallback?

One other concern I have with the code as written is if the node itself has a ContainerUser or ContainerAdministrator user, we won't actually use it even when specified.

Reply:
I think having a single documented fallback user that we use (potentially with a config option to change this) is a good idea. I think using the shim's token is unintuitive, as the shim runs as part of container infrastructure, so it would be weird if a change to its security context caused a container workload to stop working.

Reply:
Ok so the config option you were proposing was just for the default case, we're on the same page (I hope haha). I agree with having a config option/annotation to change whatever default we come up with.  

I think if the user they're asking to run as doesn't exist we should just error out instead of falling back. The user was likely chosen for a specific reason (lowered permissions, no access to certain files, etc.) so it doesn't seem right to fallback to potentially a higher privileged user (or really any other user). 

Agreed on single documented user, let me see if there's any case where Administrator wouldn't exist

Reply:
Right, but we can't always tell if they actually requested a user explicitly, right? I assume the reason this code path exists in the first place is for if no user is specified for the container, but the image defaults to ContainerUser/ContainerAdministrator.

Reply:
Yea.. that's tricky the more I think of it. How do we get around telling the difference between someone using a base image and having the user passed to us always be ContainerUser/Administrator vs. someone actually asking to run as a user with this name on the node.

Reply:
Well, never mind. We can just check if it exists on the host, and if it does use it and if not go to the fallback

Reply:
But do you agree for any other case besides ContainerUser/Admin/no user specified we shouldn't fallback if the user can't be found?

Reply:
Yes, we definitely should error if we think the user has requested a user that doesn't exist.

Reply:
If we end up creating a slim base image for privileged containers perhaps that helps solve this problem. Then we could make sure it has Administrator set in its config, and say if you use a standard image for a privileged container you need to specify a real user on the host or it won't work.

Reply:
Again, though, did we talk about this in the KEP? It seems this needs buy-in on our approach as it affects the user experience.

Reply:
We didn't mention any specific user these would run as or any expected behavior here afaik so I think we're free to make a call here

Reply:
Don't know why this didn't come to mind, but how could we use Administrator as the default if we don't have the password?

Reply:
And the other side to this, if we want to use LOCAL SERVICE/NETWORK SERVICE/SYSTEM and not just inherit from the shims token, then we need to impersonate a token from some service that's running as SYSTEM/whatever we want to run as I'm pretty sure

Reply:
I guess we don't have a way to create a proper token for this? We have to take it from an existing context?

Reply:
(or have a password)

Reply:
That's what it seems like. I'm not the most knowledgeable person here by any means so the following should be taken with a grain of salt but I was looking throughout the day to see how to get around this and it seems either, like you said, us having to have the pswd for the Administrator account if this is the route we'd want to go or somehow grab the token from another context. From digging a bit more I don't think the reverse of impersonating one of the service accounts to get a token works, I think *from* a service you're able to impersonate a token from a user account but not vice versa.

Reply:
If we don't have any viable option to make the default something other than the shim user, can we add a test that validates the user we run as is SYSTEM? That way at least if we change it we will get a clear signal that we could be breaking something.

Reply:
Yea that's a good idea. I'll make a work item for this.

Reply:
and for the launching the process *in* the job at startup

Reply:
I'd prefer if we can add the test in this PR. Otherwise it might never get done :)

Reply:
For sure, was gonna wait until we had the cri changes to add the tests but I guess I can make a work item to track changing TO the new cri changes in the test(s)